### ISSUE TITLE
FXForm and bean validation in OSGI environment

### ISSUE THREAD

**Issue by lanouettea** [dooApp/FXForm2#93]
I'm using FXForms 2 in a OSGI environment to generate forms. Classloader are very restrictive in OSGI, and the standard way of finding a validator instance in the classpath simply does not work and will give the error:

javax.validation.ValidationException: Unable to create a Configuration, because no Bean Validation provider could be found. Add a provider like Hibernate Validator (RI) to your classpath.

This is expected, and is OK. I worked around this by implementing my own FXFormValidator, extending FXForm and setting my validator in my extended class constructor. Now I can successfully validate beans in my FX Forms.

However, when I create my extended FXForm, the AbstractFXForm get instanciated. It contains the fxFormValidator member, which gets instanciated to a DefaultFXFormValidator. When the DefaultFXFormValidator gets created, it looks for a validator instance in the classpath, then cannot find it and log the stacktrace (line 54 of DefaultFXFormValidator).

This causes my logs to get filled with stacktraces everytime I create a form, before I override the form validator to my own. This is problematic and could be fixed by either:

1 - removing the log entry from the DefaultFXFormValidator;
2 - Do not instanciate the form validator at form instance creation, but only the first time a validation needs to be executed.

**Comment by amischler** [comment]
I think it's better to keep this log entry in the DefaultFXFormValidator, it might help users to understand why the validation doesn't work.

I have modified the pattern used to instantiate most properties in AbstractFXForm, so that it is easier to override the instantiation of each property.

Just override the createFxFormValidatorProperty in your case.

To make this even better, if you wish, you could also share your OSGI FormValidator to integrate it into FXForm.

**Comment by amischler** [comment]
Deployed in 8.0.8-SNAPSHOT

**Comment by lanouettea** [comment]
Thank you for the modification. Unfortunately I'm tied up on multiple things at the moment and will only be able to test around next week.

**Comment by lanouettea** [comment]
I've overriden the createFxFormValidatorProperty in my class extending FXForm.

However, I still see the issue, because this line in the AbstractFXForm constructor:

**Comment by lanouettea** [comment]
I agree that the system with factories is a good way to go for this. Unfortunately, there is a small flaw in the current 8.0.8-SNAPSHOT, which causes the code to act in a similar way as before. In AbtractFXForm.java, you eventually call the FormValidatorFactory in this section of code:

```java
classLevelValidator.validatorProperty().bind(fxFormValidatorProperty());
        classLevelValidator.constraintViolationsProperty().addListener(new ListChangeListener<ConstraintViolation>() {
@Override
     public void onChanged(Change<? extends ConstraintViolation> change) {
          while (change.next()) {
              constraintViolationsList.removeAll(change.getRemoved());
              constraintViolationsList.addAll(change.getAddedSubList());
          }
     }
});
```
This causes the defaultFXFormValidator to eventually be created in the AbstractFXForm instanciation for any case.

**Comment by amischler** [comment]
Right, the overriden method should not be called in parent constructor. Now using lazy instantiation for the classLevelValidator.

**Comment by amischler** [comment]
Fixed in 8.0.8-SNAPSHOT

**Comment by amischler** [comment]
Fixed in 8.0.8
